src/external/song.ts

```diff
@@ -17,100 +17,186 @@
+ * analyses one page at a time. The size each page is determined through the
+ * datacenter.songPageSize configuration variable.
+ *
+ * This function returns an AsyncGenerator that will yield an SongAnalyses for
+ * every page that is fetched from Song
+ *
+ * @example
+ * const analysesResponseGenerator = getAnalysesByStudy({ dataCenterId, studyId });
+ * for await (const analyses of analysesResponseGenerator) {
+ * 	// analyses contains an array of SongAnalyses objects retrieved from the specified dataCenter
+ * }
  */
-export const getAnalysesByStudy = async (url: string, studyId: string): Promise<Batch> => {
-  const analysesUrl = urljoin(url, '/studies', studyId, '/analysis?analysisStates=PUBLISHED');
-  return fetchAnalysesInBatches(analysesUrl);
-};
+export async function* getAnalysesByStudy(details: {
+	dataCenterId: string;
+	studyId: string;
+}): AsyncGenerator<SongAnalysis[]> {
+	try {
+		const dataCenter = await getDataCenter(details.dataCenterId);
+
+		const {
+			datacenter: { songPageSize },
+		} = await getAppConfig();
+
+		let allPagesFetched = false;
+		let offset = 0;
+		while (!allPagesFetched) {
+			const response = await fetchAnalysesPage({
+				dataCenter,
+				studyId: details.studyId,
+				offset,
+				limit: songPageSize,
+			});
+
+			yield response.data;
+
+			if (response.responseCount === 0 || response.responseCount < songPageSize) {
+				allPagesFetched = true;
+			} else {
+				offset += songPageSize;
+			}
+		}
+	} catch (e) {
+		logger.error('Error while fetching analyses for study');
+		logger.error(e);
+	}
+}
 
-export const getAnalysesById = async (
-  dataCenterId: string,
-  studyId: string,
-  analysisId: string,
-): Promise<SongAnalysis> => {
-  const dataCenter = await getDataCenter(dataCenterId);
-  const analysesUrl = urljoin(
-    dataCenter.songUrl,
-    '/studies',
-    studyId,
-    '/analysis',
-    analysisId,
-    '?analysisStates=PUBLISHED,UNPUBLISHED,SUPPRESSED',
-  );
-  try {
-    const res = await fetch(analysesUrl);
-    if (res.status === 200) {
-      return (await res.json()) as SongAnalysis;
-    } else {
-      logger.error(`Failure to fetch analysis ${analysisId} for ${studyId} from ${analysesUrl}`);
-      throw new Error(`Unable to retrieve analysis ${analysisId} for ${studyId} from ${analysesUrl}`);
-    }
-  } catch (e) {
-    logger.error(`Error fetching analysis ${analysisId} for ${studyId} from ${analysesUrl}: ${e}`);
-    throw new Error(`Unable to retrieve analysis ${analysisId} for ${studyId} from ${analysesUrl}`);
-  }
+export const getAnalysesById = async (inputs: {
+	dataCenterId: string;
+	studyId: string;
+	analysisId: string;
+}): Promise<SongAnalysis> => {
+	const { dataCenterId, studyId, analysisId } = inputs;
+	const dataCenter = await getDataCenter(dataCenterId);
+	const analysesUrl = urljoin(
+		dataCenter.songUrl,
+		'/studies',
+		studyId,
+		'/analysis',
+		analysisId,
+		'?analysisStates=PUBLISHED,UNPUBLISHED,SUPPRESSED',
```

Comment:
👍 nice

Reply:
Yes, but this code was not modified... simply spaces turned into tabs.